Dockerfile

```diff
@@ -29,9 +29,11 @@ RUN apt-get update \
   && rm -rf /var/lib/apt/lists/* \
   && localedef -i en_US -f UTF-8 en_US.UTF-8 \
   && useradd -m -s /bin/bash linuxbrew \
-  && echo 'linuxbrew ALL=(ALL) NOPASSWD:ALL' >>/etc/sudoers
+  && echo 'linuxbrew ALL=(ALL) NOPASSWD:ALL' >>/etc/sudoers \
+  && su - linuxbrew -c 'mkdir ~/.linuxbrew'
```

Comment:
Is this line really needed? Does COPY not make all needed dirs?

Reply:
Correct! `COPY` will not currently apply the ownership specification to anything except the last object in the path.

Reply:
_Related: I am going to be submitting a patch to `moby/moby` in the next week to resolve this_

Reply:
Can you tag me in the PR? I'd like to see it.

Reply:
Ended up getting derailed by responsibilities at `$WORK`. Will be drawing up a test case and opening an issue in `moby/moby` later today to check with the team and see if this is something they would be alright accepting a patch for, or if it is by design.

Reply:
Actually, I couldn't recreate this.

```
➜ cat Dockerfile
FROM busybox:latest
RUN adduser -D example-user
COPY --chown=example-user:example-user README /some/nested/path/README
RUN find /some -exec printf '{} | ' \; -exec stat -c 'u:%U g:%G' {} \;


➜ docker build .
Sending build context to Docker daemon  3.072kB
Step 1/4 : FROM busybox:latest
 ---> 42b97d3c2ae9
Step 2/4 : RUN adduser -D example-user
 ---> Running in bf68d8debcdf
Removing intermediate container bf68d8debcdf
 ---> e5d6c10964e7
Step 3/4 : COPY --chown=example-user:example-user README /some/nested/path/README
 ---> 08991fcbb80f
Step 4/4 : RUN find /some -exec printf '{} | ' \; -exec stat -c 'u:%U g:%G' {} \;
 ---> Running in b1b8ffbf9e5e
/some | u:example-user g:example-user
/some/nested | u:example-user g:example-user
/some/nested/path | u:example-user g:example-user
/some/nested/path/README | u:example-user g:example-user
Removing intermediate container b1b8ffbf9e5e
 ---> 4409ad3b2f87
Successfully built 4409ad3b2f87
```

So the issue I encoutered when refactoring the `Dockerfile` in this repository is likely caused by something else.

Reply:
With an example more aligned with this repository's Dockerfile:

```
➜ docker build --force-rm .
Sending build context to Docker daemon  3.072kB
Step 1/5 : FROM busybox:latest
 ---> 42b97d3c2ae9
Step 2/5 : RUN adduser -D example-user
 ---> Using cache
 ---> e5d6c10964e7
Step 3/5 : USER example-user
 ---> Running in 059d30c97b5e
Removing intermediate container 059d30c97b5e
 ---> 7a11207de194
Step 4/5 : COPY --chown=example-user:example-user README /home/example-user/.example-user/README
 ---> 621af6a731f6
Step 5/5 : RUN find /home -exec printf '{} | ' \; -exec stat -c 'u:%U g:%G' {} \;
 ---> Running in d41f0dbca6cd
/home | u:nobody g:nobody
/home/example-user | u:example-user g:example-user
/home/example-user/.example-user | u:example-user g:example-user
/home/example-user/.example-user/README | u:example-user g:example-user
Removing intermediate container d41f0dbca6cd
 ---> 2791646ab0d9
Successfully built 2791646ab0d9
```

Reply:
Aha! It occurs when copying directories:

```
➜ docker build --force-rm .
Sending build context to Docker daemon  3.584kB
Step 1/5 : FROM busybox:latest
 ---> 42b97d3c2ae9
Step 2/5 : RUN adduser -D example-user
 ---> Using cache
 ---> e5d6c10964e7
Step 3/5 : USER example-user
 ---> Using cache
 ---> 7a11207de194
Step 4/5 : COPY --chown=example-user:example-user somedir /home/example-user/.example-user/somedir
 ---> 42c6b01a20ef
Step 5/5 : RUN find /home -exec printf '{} | ' \; -exec stat -c 'u:%U g:%G' {} \;
 ---> Running in 8d636009ba0f
/home | u:nobody g:nobody
/home/example-user | u:example-user g:example-user
/home/example-user/.example-user | u:root g:root
/home/example-user/.example-user/somedir | u:example-user g:example-user
/home/example-user/.example-user/somedir/README | u:example-user g:example-user
Removing intermediate container 8d636009ba0f
 ---> f6837be02431
Successfully built f6837be02431
```

Reply:
Slimmed down test (removed the `USER` instruction):

```
➜ docker build --force-rm .
Sending build context to Docker daemon  3.584kB
Step 1/4 : FROM busybox:latest
 ---> 42b97d3c2ae9
Step 2/4 : RUN adduser -D example-user
 ---> Using cache
 ---> 5b53d5285fc1
Step 3/4 : COPY --chown=example-user:example-user bar /home/example-user/foo/bar
 ---> 7d707703388b
Step 4/4 : RUN find /home -exec printf '{} | ' \; -exec stat -c 'u:%U g:%G' {} \;
 ---> Running in c044b0ad97fb
/home | u:nobody g:nobody
/home/example-user | u:example-user g:example-user
/home/example-user/foo | u:root g:root
/home/example-user/foo/bar | u:example-user g:example-user
/home/example-user/foo/bar/README | u:example-user g:example-user
Removing intermediate container c044b0ad97fb
 ---> 6bb7ea11db88
Successfully built 6bb7ea11db88
```

I've opened https://github.com/moby/moby/issues/42819.